### PR TITLE
Prepare for Vapor 3

### PR DESCRIPTION
**Pull request by BrettRToomey**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -2,18 +2,20 @@ version: 2
 jobs:
   MacOS:
     macos:
-      xcode: "9.0"
+      xcode: "9.3.0"
```
Comment (steffendsommer) [inline .circleci/config.yml]:
```suggestion
      xcode: "10.0.0"
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -2,18 +2,20 @@ version: 2
...
     steps:
       - checkout
       - restore_cache:
           keys:
-            - v1-spm-deps-{{ checksum "Package.swift" }}
+            - v3-spm-deps-{{ checksum "Package.swift" }}
```
Comment (steffendsommer) [inline .circleci/config.yml]:
```suggestion
            - v1-spm-deps-{{ checksum "Package.swift" }}
```

Updating `Package.swift` to 4.2 should allow us to go back to v1 and v2.

File: .circleci/config.yml [hunk 3]
```diff
@@ -25,17 +27,17 @@ jobs:
           command: |
             bash <(curl -s https://codecov.io/bash)
       - save_cache:
-          key: v1-spm-deps-{{ checksum "Package.swift" }}
+          key: v3-spm-deps-{{ checksum "Package.swift" }}
```
Comment (steffendsommer) [inline .circleci/config.yml]:
```suggestion
          key: v2-spm-deps-{{ checksum "Package.swift" }}
```

File: .circleci/config.yml [hunk 4]
```diff
@@ -25,17 +27,17 @@ jobs:
...
           paths:
             - .build
   Linux:
     docker:
-      - image: brettrtoomey/vapor-ci:0.0.1
+      - image: nodesvapor/vapor-ci:swift-4.1
```
Comment (steffendsommer) [inline .circleci/config.yml]:
```suggestion
      - image: nodesvapor/vapor-ci:swift-4.2
```

File: .circleci/config.yml [hunk 5]
```diff
@@ -25,17 +27,17 @@ jobs:
...
     steps:
       - checkout
       - restore_cache:
           keys:
-            - v2-spm-deps-{{ checksum "Package.swift" }}
+            - v4-spm-deps-{{ checksum "Package.swift" }}
```
Comment (steffendsommer) [inline .circleci/config.yml]:
```suggestion
            - v2-spm-deps-{{ checksum "Package.swift" }}
```

File: .circleci/config.yml [hunk 6]
```diff
@@ -48,7 +50,7 @@ jobs:
           name: Restoring Package file
           command: mv res Package.swift
       - save_cache:
-          key: v2-spm-deps-{{ checksum "Package.swift" }}
+          key: v4-spm-deps-{{ checksum "Package.swift" }}
```
Comment (steffendsommer) [inline .circleci/config.yml]:
```suggestion
          key: v2-spm-deps-{{ checksum "Package.swift" }}
```

File: Package.swift [hunk 7]
```diff
@@ -1,11 +1,27 @@
+// swift-tools-version:4.0
```
Comment (steffendsommer) [inline Package.swift]:
```suggestion
// swift-tools-version:4.2
```

File: README.md [hunk 8]
```diff
@@ -1,6 +1,6 @@
 # Storage 🗄
-[![Swift Version](https://img.shields.io/badge/Swift-3-brightgreen.svg)](http://swift.org)
-[![Vapor Version](https://img.shields.io/badge/Vapor-2-F6CBCA.svg)](http://vapor.codes)
+[![Swift Version](https://img.shields.io/badge/Swift-4-brightgreen.svg)](http://swift.org)
```
Comment (steffendsommer) [inline README.md]:
```suggestion
[![Swift Version](https://img.shields.io/badge/Swift-4.2-brightgreen.svg)](http://swift.org)
```

File: README.md [hunk 9]
```diff
@@ -1,6 +1,6 @@
...
+[![Vapor Version](https://img.shields.io/badge/Vapor-3-F6CBCA.svg)](http://vapor.codes)
```
Comment (steffendsommer) [inline README.md]:
```suggestion
[![Vapor Version](https://img.shields.io/badge/Vapor-3-30B6FC.svg)](http://vapor.codes)
```

File: Sources/Storage/NetworkDriver.swift [hunk 10]
```diff
@@ -53,17 +65,22 @@ final class S3Driver: NetworkDriver {
...
+            on: container
+        ).map { _ in
+            return path
+        }
     }
     
-    func get(path: String) throws -> Bytes {
-        return try s3.get(path: path)
+    public func get(path: String, on container: Container) throws -> Future<[UInt8]> {
+        return container.future([])
```
Comment (steffendsommer) [inline Sources/Storage/NetworkDriver.swift]: Should this be removed and added as issues so we can keep track of missing features?

File: Sources/Storage/NetworkDriver.swift [hunk 11]
```diff
@@ -53,17 +65,22 @@ final class S3Driver: NetworkDriver {
...
     }
     
-    func delete(path: String) throws {
-        return try s3.delete(file: path)
+    public func delete(path: String, on container: Container) throws -> Future<Void> {
+        return container.future()
```
Comment (steffendsommer) [inline Sources/Storage/NetworkDriver.swift]: Should this be removed and added as issues so we can keep track of missing features?

File: Sources/Storage/Utilities/DataURI.swift [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+     - Returns: (data: [UInt8], type: [UInt8], typeMetadata: [UInt8]?)
+     */
+    public static func parse(uri: String) throws -> (Data, [UInt8], [UInt8]?) {
+        guard uri.hasPrefix("data:") else {
+            throw Error.invalidScheme
+        }
+
+        var scanner: Scanner<UInt8> = Scanner(uri.bytes)
+        //pop scheme ("data:")
```
Comment (steffendsommer) [inline Sources/Storage/Utilities/DataURI.swift]: Can this comment be removed?

File: Sources/Storage/Utilities/S3.swift [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+    func hashed() throws -> String {
+        switch self {
+        case .bytes(let bytes):
+            return try SHA256.hash(bytes).hexEncodedString()
+
+        case .unsigned:
+            return "UNSIGNED-PAYLOAD"
+
+        case .none:
+            return "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
```
Comment (steffendsommer) [inline Sources/Storage/Utilities/S3.swift]: What is this? Consider adding a comment explaining it

File: Sources/Storage/Utilities/S3.swift [hunk 14]
```diff
@@ -0,0 +1,362 @@
...
+        return signature.hexEncodedString()
+    }
+
+    func getCredentialScope() -> String {
+        return [
+            dateStamp(),
+            region,
+            service,
+            "aws4_request"
+            ].joined(separator: "/")
```
Comment (steffendsommer) [inline Sources/Storage/Utilities/S3.swift]: indentation

File: Sources/Storage/Utilities/S3.swift [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+        date: String,
+        scope: String,
+        canonicalHash: String
+        ) -> String {
+        return [
+            algorithm,
+            date,
+            scope,
+            canonicalHash
+            ].joined(separator: "\n")
```
Comment (steffendsommer) [inline Sources/Storage/Utilities/S3.swift]: indentation

File: Sources/Storage/Utilities/S3.swift [hunk 16]
```diff
@@ -0,0 +1,362 @@
...
+        return dateFormatter.string(from: unitTestDate ?? Date())
+    }
+
+    public init(
+        service: String,
+        host: String,
+        region: Region,
+        accessKey: String,
+        secretKey: String
+        ) {
```
Comment (steffendsommer) [inline Sources/Storage/Utilities/S3.swift]: indentation

File: Sources/Storage/Utilities/S3.swift [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+    }
+
+    func getCanonicalRequest(
+        payloadHash: String,
+        method: Method,
+        path: String,
+        query: String,
+        canonicalHeaders: String,
+        signedHeaders: String
+        ) throws -> String {
```
Comment (steffendsommer) [inline Sources/Storage/Utilities/S3.swift]: indentation

File: Sources/Storage/Utilities/S3.swift [hunk 18]
```diff
@@ -0,0 +1,362 @@
...
+        return dateFormatter.string(from: date)
+    }
+}
+
+extension AWSSignatureV4 {
+    func generateHeadersToSign(
+        headers: inout [String: String],
+        host: String,
+        hash: String
+        ) {
```
Comment (steffendsommer) [inline Sources/Storage/Utilities/S3.swift]: indentation

File: Sources/Storage/Utilities/S3.swift [hunk 19]
```diff
@@ -0,0 +1,362 @@
...
+            "\($0.key.lowercased()):\($0.value)"
+            }.joined(separator: "\n")
+    }
+
+    func createAuthorizationHeader(
+        algorithm: String,
+        credentialScope: String,
+        signature: String,
+        signedHeaders: String
+        ) -> String {
```
Comment (steffendsommer) [inline Sources/Storage/Utilities/S3.swift]: indentation

File: Sources/Storage/Utilities/S3.swift [hunk 20]
```diff
@@ -0,0 +1,362 @@
...
+    }
+
+    func alphabetize(_ dict: [String : String]) -> [(key: String, value: String)] {
+        return dict.sorted(by: { $0.0.lowercased() < $1.0.lowercased() })
+    }
+
+    func createCanonicalHeaders(_ headers: [(key: String, value: String)]) -> String {
+        return headers.map {
+            "\($0.key.lowercased()):\($0.value)"
+            }.joined(separator: "\n")
```
Comment (steffendsommer) [inline Sources/Storage/Utilities/S3.swift]: indentation

File: Sources/Storage/Utilities/S3.swift [hunk 21]
```diff
@@ -0,0 +1,362 @@
...
+     - query: Additional querystring in key-value format ("?key=value&key2=value2")
+     - headers: HTTP headers added to the request
+     */
+    public func sign(
+        payload: Payload = .none,
+        method: Method = .get,
+        path: String,
+        query: String? = nil,
+        headers: [String : String] = [:]
+        ) throws -> [String : String] {
```
Comment (steffendsommer) [inline Sources/Storage/Utilities/S3.swift]: indentation

File: Sources/Storage/Utilities/S3.swift [hunk 22]
```diff
@@ -0,0 +1,362 @@
...
+        self.accessKey = accessKey
+        self.secretKey = secretKey
+    }
+
+    func getStringToSign(
+        algorithm: String,
+        date: String,
+        scope: String,
+        canonicalHash: String
+        ) -> String {
```
Comment (steffendsommer) [inline Sources/Storage/Utilities/S3.swift]: indentation

File: Sources/Storage/Utilities/S3.swift (hunk 13 above)
Comment (BrettRToomey) [reply Sources/Storage/Utilities/S3.swift]: Nothing hashed

File: Sources/Storage/Utilities/DataURI.swift (hunk 12 above)
Comment (BrettRToomey) [reply Sources/Storage/Utilities/DataURI.swift]: Sure, but it shows where the number in the following line comes from

File: Sources/Storage/NetworkDriver.swift (hunk 10 above)
Comment (BrettRToomey) [reply Sources/Storage/NetworkDriver.swift]: Good call

File: Sources/Storage/NetworkDriver.swift (hunk 11 above)
Comment (BrettRToomey) [reply Sources/Storage/NetworkDriver.swift]: Good call

**Pull request merged**